litmus-portal/graphql-server/pkg/myhub/handler/handler.go

```diff
@@ -191,3 +198,139 @@ func IsFileExisting(path string) (bool, error) {
 	}
 	return true, nil
 }
+
+//DownloadRemoteHub is used to download a remote hub from the url provided by the user
+func DownloadRemoteHub(hubDetails model.CreateRemoteMyHub) error {
+	//create the destination directory where the hub will be downloaded
+	hubpath := defaultPath + hubDetails.ProjectID + "/" + hubDetails.HubName + ".zip"
+	destDir, err := os.Create(hubpath)
+	if err != nil {
+		fmt.Println(err)
+		return err
+	}
+	defer destDir.Close()
+
+	//download the zip file from the provided url
+	download, err := http.Get(hubDetails.RepoURL)
+	if err != nil {
+		fmt.Println(err)
+		return err
+	}
+	if download.StatusCode != http.StatusOK {
+		err = fmt.Errorf("err: ", download.Status)
+		return err
+	}
+
+	//validate the content length (in bytes)
+	maxSize, err := strconv.Atoi(os.Getenv("REMOTE_HUB_MAX_SIZE"))
+	if err != nil {
+		return err
+	}
+	contentLength := download.Header.Get("content-length")
+	length, err := strconv.Atoi(contentLength)
+	if length > maxSize {
+		_ = os.Remove(hubpath)
+		err = fmt.Errorf("err: File size exceeded the threshold %d", length)
+		return err
+	}
```

Comment:
If we need to validate file size it should be done before downloading the actual file, if the file size is huge it can cause unwanted crashes as we blindly download the file. Though we do not bother with repo size when we use normal git so this might be a disparity between the 2 features. A large git repo might be cloned directly but it may fail when using the remote download feature.

Reply:
```
         resp, err := http.Head(url)
         if err != nil {
               // handle error
         }

         if resp.StatusCode != http.StatusOK {
                 // handle error
         }

         // check "Content-Length" for file size
         size, err := strconv.Atoi(resp.Header.Get("Content-Length"))
          if err != nil {
               // handle error
         }
         downloadSize := int64(size)
```

Reply:
try this ^^

Reply:
Thanks for the review @gdsoumya 
The file will be copied after all the following checks https://github.com/litmuschaos/litmus/blob/02837869e58415c27523deaa374a4fd071e7b9a1/litmus-portal/graphql-server/pkg/myhub/handler/handler.go#L248

If we use http.Head first to validate the file type and size, we will have to do another http.Get to fetch the body/data. I'm not sure we need two http calls here.

Reply:
Okay, maybe then move up the resp.Body.Close() to the top because the buffer needs to be closed in case the func errors out before the body is read.

Reply:
Thanks @gdsoumya ! Have made the change! 😄